File: homework03/life.py
import pathlib
import random

from typing import List, Optional, Tuple

Cell = Tuple[int, int]
Cells = List[int]
Grid = List[Cells]


class GameOfLife:

    def __init__(
            self,
            size: Tuple[int, int],
            randomize: bool = True,
            max_generations: Optional[float] = float('inf')
    ) -> None:
        # Размер клеточного поля
        self.rows, self.cols = size
        # Предыдущее поколение клеток
        self.prev_generation = self.create_grid()
        # Текущее поколение клеток
        self.curr_generation = self.create_grid(randomize=randomize)
        # Максимальное число поколений
        self.max_generations = max_generations
        # Текущее число поколений
        self.generations = 1

    def create_grid(self, randomize: bool = False) -> Grid:
        grid = [[0] * self.cols for _ in range(self.rows)]
        if randomize:
            for i in range(self.rows):
                for j in range(self.cols):
                    grid[i][j] = random.randint(0, 1)
        return grid

    def get_neighbours(self, cell: Cell) -> Cells:
        row, col = cell
        res = []
        if row > 0:
            res.append(self.curr_generation[row - 1][col])
            if col > 0:
                res.append(self.curr_generation[row - 1][col - 1])
            if col < self.cols - 1:
                res.append(self.curr_generation[row - 1][col + 1])

        if row < self.rows - 1:
            res.append(self.curr_generation[row + 1][col])
            if col > 0:
                res.append(self.curr_generation[row + 1][col - 1])
            if col < self.cols - 1:
                res.append(self.curr_generation[row + 1][col + 1])

        if col > 0:
            res.append(self.curr_generation[row][col - 1])
        if col < self.cols - 1:
            res.append(self.curr_generation[row][col + 1])

        return res

    def get_next_generation(self) -> Grid:
        for i in range(self.rows):
            for j in range(self.cols):
                self.prev_generation[i][j] = self.curr_generation[i][j]

        alive = []
        dead = []

        for i in range(self.rows):
            for j in range(self.cols):
                neighbours = sum(self.get_neighbours((i, j)))
                if neighbours < 2 or neighbours > 3:
                    dead.append((i, j))
                elif neighbours == 3 and self.curr_generation[i][j] == 0:
                    alive.append((i, j))

        for t in alive:
            self.curr_generation[t[0]][t[1]] = 1
        for t in dead:
            self.curr_generation[t[0]][t[1]] = 0
        return self.curr_generation

    def step(self) -> None:
        """
        Выполнить один шаг игры.
        """
        self.get_next_generation()
        self.generations += 1

    @property
    def is_max_generations_exceeded(self) -> bool:
        """
        Не превысило ли текущее число поколений максимально допустимое.
        """
        return self.generations >= self.max_generations

    @property
    def is_changing(self) -> bool:
        """
        Изменилось ли состояние клеток с предыдущего шага.
        """
        for i in range(self.rows):
            for j in range(self.cols):
                if self.prev_generation[i][j] != self.curr_generation[i][j]:
                    return True
        return False

    @staticmethod
    def from_file(filename: pathlib.Path) -> 'GameOfLife':
        """
        Прочитать состояние клеток из указанного файла.
        """

        with open(filename, 'r') as f:
            lines = f.readlines()
        rows = len(lines)
        cols = len(lines[0].strip())
        game = GameOfLife(size=(rows, cols), randomize=False)

        with open(filename, 'r') as f:
            for i in range(rows):
                lines[i].strip()
                for j in range(cols):
                    game.curr_generation[i][j] = int(lines[i][j])
        return game

    def save(self, filename: pathlib.Path) -> None:
        """
        Сохранить текущее состояние клеток в указанный файл.
        """
        with open(filename, 'w') as f:
            for row in self.curr_generation:
                f.write(''.join(str(v) for v in row) + '\n')
        return

File: homework03-go/life/life.go
package life

import (
	"fmt"
  "os"
  "strconv"
  "strings"
  "bufio"
	"math/rand"
)

type GameOfLife struct {
	rows, cols, generations, max_generations int
	prev_generation, curr_generation         [][]int
}

type Cell struct {
	row, col int
}

func newGame(rows int, cols int, randomize bool, max_generations int) GameOfLife {
	prev_generation := CreateGrid(rows, cols, false)
	curr_generation := CreateGrid(rows, cols, randomize)
	return GameOfLife{
		rows:            rows,
		cols:            cols,
		prev_generation: prev_generation,
		curr_generation: curr_generation,
		max_generations: max_generations,
		generations:     1,
	}
}

func CreateGrid(rows int, cols int, randomize bool) [][]int {
	grid := make([][]int, rows)

	for i := 0; i < rows; i++ {
		grid[i] = make([]int, cols)
		if randomize == true {
			for j := 0; j < cols; j++ {
				grid[i][j] = rand.Intn(2)
			}
		}
	}
	return grid
}

func (game GameOfLife) GetNeighbours(cell Cell) []int {
	row, col := cell.row, cell.col
	res := make([]int, 0)

	if row > 0 {
		res = append(res, game.curr_generation[row-1][col])
		if col > 0 {
			res = append(res, game.curr_generation[row-1][col-1])
		}
		if col < game.cols-1 {
			res = append(res, game.curr_generation[row-1][col+1])
		}
	}
	if row < game.rows-1 {
		res = append(res, game.curr_generation[row+1][col])
		if col > 0 {
			res = append(res, game.curr_generation[row+1][col-1])
		}
		if col < game.cols-1 {
			res = append(res, game.curr_generation[row+1][col+1])
		}
	}
	if col > 0 {
		res = append(res, game.curr_generation[row][col-1])
	}
	if col < game.cols-1 {
		res = append(res, game.curr_generation[row][col+1])
	}
	return res
}

func (game *GameOfLife) GetNextGeneration() [][]int {
	for i := 0; i < game.rows; i++ {
		for j := 0; j < game.cols; j++ {
			game.prev_generation[i][j] = game.curr_generation[i][j]
		}
	}

	alive := make([]Cell, 0)
	dead := make([]Cell, 0)
	for i := 0; i < game.rows; i++ {
		for j := 0; j < game.cols; j++ {
			neighbours := game.GetNeighbours(Cell{i, j})
			sum := 0
			for _, n := range neighbours {
				sum += n
			}

			if sum < 2 || sum > 3 {
				dead = append(dead, Cell{i, j})
			} else if sum == 3 && game.curr_generation[i][j] == 0 {
				alive = append(alive, Cell{i, j})
			}
		}
	}

	for _, a := range alive {
		game.curr_generation[a.row][a.col] = 1
	}
	for _, d := range dead {
		game.curr_generation[d.row][d.col] = 0
	}
	return game.curr_generation
}

func (game *GameOfLife) Step() {
	game.GetNextGeneration()
	game.generations++
}

func (game *GameOfLife) Is_Max_Generations_Exceeded() bool {
	return game.generations >= game.max_generations
}

func (game *GameOfLife) Is_Changing() bool {
	for i := 0; i < game.rows; i++ {
		for j := 0; j < game.cols; j++ {
			if game.curr_generation[i][j] != game.prev_generation[i][j] {
				return true
			}
		}
	}
	return false
}

func FromFile(filename string) GameOfLife {
  file, err := os.Open(filename) 
    if err != nil { 
      fmt.Println(err) 
      os.Exit(1)
    } 
  defer file.Close() 
  
  reader := bufio.NewReader(file)
  var grid [][]int

  for { 
    line, err := reader.ReadString('\n') 
    if err != nil {
      break
    }
    str_arr := strings.Split(strings.Trim(line, "\n"), "")
    var int_arr []int
    for _, str := range str_arr {
      temp, _ := strconv.ParseInt(str, 10, 64)
      int_arr = append(int_arr, int(temp))
    }
    grid = append(grid, int_arr)
  }

  game := newGame(len(grid), len(grid[0]), false, 1000)
  game.curr_generation = grid
  return game
}

func (game *GameOfLife) Save (filename string) {
  file, err := os.Create(filename) 
  writer := bufio.NewWriter(file) 
  if err != nil { 
    fmt.Println(err) 
    os.Exit(1) 
  } 
  defer file.Close() 
  
  for _, row := range game.curr_generation {
    var str_res string
    for _, v := range row {
      str_res += strconv.Itoa(v)
    }
    writer.WriteString(str_res + "\n")
  } 
  writer.Flush() 
}

func Main() {
	//game := FromFile("life/grid.txt")
	game := newGame(5, 5, true, 5)

	for i := 0; i < 4; i++ {
	  game.Step()
	  fmt.Println("generations =", game.generations)
	  fmt.Println("max_generations =", game.max_generations)
	  fmt.Println("exceeded =", game.Is_Max_Generations_Exceeded())
	  fmt.Println("changing =", game.Is_Changing())
	  fmt.Println(game.prev_generation)
	  fmt.Println()
	}
  game.Save("life/curr_gen.txt")
}
